projects/igniteui-angular/src/lib/time-picker/time-picker.component.ts

```diff
@@ -1073,6 +1326,240 @@ export class IgxTimePickerComponent implements
+                break;
+            case 'arrowdown':
+            case 'down':
+                if (alt) {
+                    this.openDialog();
+                } else {
+                    this.spinOnEdit(event);
+                }
+                break;
+            default:
+                return;
+        }
+    }
+
+    /**
+     * @hidden
+     */
+    public onInput(event): void {
+        const val = event.target.value;
+        const oldVal = this.value;
+        let newVal = this._convertMinMaxValue(val);
+
+        this.isNotEmpty = val !== this.parseMask(false);
+
+        // handle cases where all empty positions (promts) are filled and we want to update
+        // timepicker own value property if it is a valid Date
+        if (val.indexOf(this.promptChar) === -1) {
+            if (this._isEntryValid(val)) {
+                if (oldVal.getTime() !== newVal.getTime()) {
+                    this.value = newVal;
+                }
+            } else {
+                const args: IgxTimePickerValidationFailedEventArgs = {
+                    timePicker: this,
+                    currentValue: val,
+                    setThroughUI: false
+                };
+                this.onValidationFailed.emit(args);
+            }
+        }
+
+        // handle cases where the user deletes the display value (when pressing backspace or delete)
+        if (!this.value || !val || val === this.parseMask(false)) {
+            this.isNotEmpty = false;
+            this.value.setHours(0, 0);
+
+            newVal = new Date(this.value);
+            if (oldVal.getTime() !== newVal.getTime()) {
+                const args: IgxTimePickerValueChangedEventArgs = {
+                    oldValue: oldVal,
+                    newValue: this.value
+                };
+                this.onValueChanged.emit(args);
+            }
+        }
+    }
+
+    /**
+     * @hidden
+     */
+    public onFocus(event): void {
+        this.isNotEmpty = event.target.value !== this.parseMask(false);
+    }
+
+    /**
+     * @hidden
+     */
+    public onBlur(event): void {
+        const value = event.target.value;
+        const oldVal = new Date(this.value);
+
+        this.isNotEmpty = value !== '';
+        this.displayValue = value;
+
+        if (value && value !== this.parseMask()) {
+            if (this._isEntryValid(value)) {
+                const newVal = this._convertMinMaxValue(value);
+                if (oldVal.getTime() !== newVal.getTime()) {
+                    this.value = newVal;
+                }
+            } else {
+                const args: IgxTimePickerValidationFailedEventArgs = {
+                    timePicker: this,
+                    currentValue: value,
+                    setThroughUI: false
+                };
+                this.onValidationFailed.emit(args);
+            }
+        }
+    }
+
+    /**
+     * @hidden
+     */
+    public spinOnEdit(event): void {
+        event.preventDefault();
```

Comment:
The complexity of this method is very high - 17, it should be 6-7. Try to refactor this and make smaller functions :)